File: 团体程序设计天梯赛-练习集/L2-014_列车调度.go
package main

import "fmt"

// case 1,2,3 timeout
func main() {
	var n int
	fmt.Scanf("%d", &n)

	s := make([]int, 0)
	for i := 0; i < n; i++ {
		var t int
		fmt.Scan(&t)
		p := UpperBound(s, t)
		if p == len(s) {
			s = append(s, t)
		} else {
			s[p] = t
		}
	}
	fmt.Println(len(s))
}

func UpperBound(arr []int, target int) int {
	lo, hi := 0, len(arr)-1
	for lo <= hi {
		mid := lo + (hi-lo)/2
		if arr[mid] > target {
			hi = mid - 1
		} else {
			lo = mid + 1
		}
	}
	return lo
}

File: 团体程序设计天梯赛-练习集/L2-014_列车调度.cpp
#include <iostream>
#include <set>
using namespace std;
/*
Principle: maintain the set of the smallest end of the trains
1. if coming train's number is larger than all exist numbers in ordered set, then add it to the set
2. else, remove next larger train's number(use lower_bound/upper_bound to find it) in set, and add the new train to the set
*/
int main() {
    ios::sync_with_stdio(false);
    int n, t;
    set<int> s;
    cin >> n;
    for (int i = 0; i < n; i++) {
        cin >> t;
        auto it = s.upper_bound(t);
        if (it == s.end()) s.insert(t); 
        else {
            s.erase(it);
            s.insert(t);
        }
    }
    cout << s.size() << endl;
    return 0;
}

/* 
//another implementation version
#include <iostream>
#include <vector>
using namespace std;
// Principle: maintain the set of the smallest end of the trains
// 1. if coming train's number is larger than all exist numbers in ordered set, then add it to the set
// 2. else, remove next larger train's number(use lower_bound/upper_bound to find it) in set, and add the new train to the set

int UpperBound(const vector<int> &arr, int target) {
    int low = 0, high = arr.size() - 1;
    while (low <= high) {
        int mid = (low + high) / 2;
        if (arr[mid] > target) high = mid - 1;
        else low = mid + 1;
    }
    return low;
}

int main() {
    ios::sync_with_stdio(false);
    int n, t;
    vector<int> s;
    cin >> n;
    for (int i = 0; i < n; i++) {
        cin >> t;
        int pos = UpperBound(s, t);
        if (pos == s.size()) s.push_back(t); 
        else s[pos] = t;
    }
    cout << s.size() << endl;
    return 0;
}
*/
